source/common/grpc/async_client.cc

```diff
@@ -0,0 +1,55 @@
+#include "envoy/grpc/async_client.h"
```

Comment:
I'm not keen to this structure, implementing interface in `include` without Impl classes. It also add an transitive dependency of the interface. 

Can we have Raw and Typed classes, but not inherit but just own Raw in Typed?

Reply:
Is this similar to what htuch@ is suggesting about having the Typed version in source/... ?  It seems like it would be possible to do what you are suggesting, although keeping the Typed version abstract at the interface level without inheritance could get tricky, e.g. requiring a factory for example which might be as complicated.   If this was just implementation then it is definitely possible.  However as I said above, it would make mocking a bit trickier and it would require dipping into the implementation in the common case.  Could give me a bit more of an idea of what you are proposing?

Reply:
Updated to raw only API.